### PR TITLE
New physics behavior

### PR DESCRIPTION
**Pull request by Lizard-13**
- [x] Previous functionality
- [x] Edge shape
- [x] Custom shape dimensions and offset
- [x] Kinematic type
- [x] Filtering layers and masks
- [x] Ability to apply forces and torques at any point (not only mass center)
- [x] Joints ID manager
- [x] Joints
  - [x] Distance joint
  - [x] Revolute joint (fixed point or between objects)
  - [x] Prismatic joint
  - [x] Pulley joint
  - [x] Gear joint
  - [x] Mouse joint
  - [x] Wheel joint
  - [x] Weld joint
  - [x] Rope joint
  - [x] Friction joint
  - [x] Motor joint
- [x] Ability to remove joints
- [x] Conditions to check joint reaction force and torque (useful to make breakable joints)
- [x] Ability to modify any joint property on the fly
- [x] When object size is modified, recreate the shape only, not the entire object, or remake the joints in the new body
- [x] Expression to get body mass center
- [x] Condition to check colliding objects
- [x] Remake the icons and add new ones

Also simplified the world scaling, no more negative values and opposite angles here and there :)

EDIT: Forget to say that almost every position and angle value in joints settings are in world coordinates, so you can add joints using sprites points for example, without worrying about the object rotation.

### CODE REVIEW

**Comment by 4ian** [comment]
Incredible 😍

**Comment by zatsme** [comment]
Loving this, but you made a **copypasta** error I think 😄 

In **physikruntimebehaviour.js**

The Wheel Joint Definition.... 

Line 1686

`gdjs.PhysikRuntimeBehavior.prototype.addWheelJoint = function(x1, y1, other, x2, y2, axisAngle, frequency, dampingRatio, enableMotor, motorSpeed, maxMotorForce, collideConnected, variable){`

and line 1711

` jointDef.set_maxMotorForce(maxMotorForce);`

Both should be **maxMotorTorque**

Apart from that I got the Wheels working fine 👍

**Comment by zatsme** [comment]
bikes

**Comment by 4ian** [comment]
That's really cool! We should definitely have both the demo that you made as example! :)

**Comment by Lizard-13** [comment]
Fixed!, thanks, surely I've a lot of these, there are tons of settings :D
I'll finish the other things and start testing _every_ setting

**Comment by zatsme** [comment]
One other thing, It took a few seconds for me to find the Mass setting.... It is labelled Density, which is fine if you know this stuff but for beginners I think either Mass (Density) or vice versa, or just Mass 🤔

**Comment by zatsme** [comment]
Ok, I'm used to setting mass explicitly, but I like the way box2d does it 😁. This is going to be super powerful 😊

**Comment by zatsme** [comment]
No, it's good, we'll make it clear in the doc 😁

**Comment by Lizard-13** [comment]
Now, on object size change, instead destroying and recreating the body the fixture is recreated, this saves the joints and I guess it's a bit faster.
You can use custom shape dimensions and offset, to achieve things like this:
shapessize
It's confusing until you can see the real shapes.
There's also an action to set the shape scale, it works when using custom shape dimensions or offsets, so you can scale it with the object.

**Comment by Lizard-13** [comment]
Ok, it should be *almost* free of bugs, tested every setting, just missed some default values on joint creation
itsalive

And here are the test files
The testbed with all the joints: 
[PhysikJointsTest.zip](https://github.com/4ian/GDevelop/files/2637436/PhysikJointsTest.zip)
All the joints with editable settings:
[PhysikJointsSettingsTest.zip](https://github.com/4ian/GDevelop/files/2637439/PhysikJointsSettingsTest.zip)
And the boring one where you can change the custom shapes scale:
[Uploading PhysikObjectSizeTest.zip…]()

**Comment by zatsme** [comment]
Looking very good 😊.

Just 1 question right now, how are you supposed to use the edge shape, and when might we get polygons? 😋

**Comment by Lizard-13** [comment]
The edge shape is almost the same than a box, but it's mean to be static and is a bit faster, so it's better suited for level grounds. 

Polygons could be already added, I did some code in a custom test file, so I know it works and can already put the working code in GD... But to be useful we need a polygon editor.
@4ian Can we create custom editors for behaviors settings?, right now the editor is generated automatically, so I can't make a better layers/masks selector or make it to preview the shape or a polygon editor. Surely it would take me a lot of time to finish something like that, but I world do it :)

**Comment by zatsme** [comment]
Can't you get the polygon data from the sprite polygon data.  We can already edit that, although it's a little clumsy interface as added dots all go to 0,0 right now.

**Comment by Lizard-13** [comment]
Although it's the most common use case, the object to which you attach the behavior is not always a sprite.
It would be super dirty, but I could enable the Polygon shape, and in the behavior code check if the owner object is a sprite, if it's, get the polygon from the collision shape from the sprite first animation and frame. Don't like it a lot.

**Comment by 4ian** [comment]
> @4ian Can we create custom editors for behaviors settings?, right now the editor is generated automatically, so I can't make a better layers/masks selector or make it to preview the shape or a polygon editor. Surely it would take me a lot of time to finish something like that, but I world do it :)

So for now the editor for behaviors (`BehaviorsEditor/index.js`) is reading the properties of the behavior and then using `PropertiesEditor` to generate a form from these properties. Precisely a schema is generated from the properties (using `PropertiesMapToSchema.js`) and sent to `PropertiesEditor.
* `PropertiesEditor` support displaying a button (search for `_renderButton` in `PropertiesEditor/index.js`) if the field of the schema has "onClick".
* We could add a new property or a special kind of property in `PropertiesMapToSchema` to handle buttons opening dialogs.
* The difficulty is to actually open the dialog: how to register/find the dialog to open? I guess we would need to "register" somewhere a list of dialog that can be opened from PropertiesEditor (much like `ParameterRenderingService`, which given a type of parameter returns the class to be used to do the rendering). We would have a `PropertyEditorService` that would be, given an editor name, open a dialog with this editor - and would have a method to register a new editor.
* The special property I mentioned before would be having the name of the editor stored, maybe in the properties "extraInfo".

Finally, the question is: where to write this new editor/dialog? For now, the answer is: directly inside newIDE. Ideally the extension would be able to bring their own editors as part of the extension source code. But this is difficult because the "compilation" of newIDE is done only for files in `newIDE/app/src/*`.

You can see that for example, for now, all objects editors are in `newIDE/app/src/ObjectEditor/Editors` (and there is a `ObjectsEditorService` for giving access to them).

So the answer would be: we need to create a `PropertyEditorService` in `PropertiesEditor` folder (and a subfolder "Editor" containing the custom editors) (like it's done for `ObjectsEditorService.js`).
(Later if we are able to compile .js file in extensions for the IDE, we will move all the files in the `**/Editors/*` folders to their own extensions).

**Comment by 4ian** [comment]
> It would be super dirty, but I could enable the Polygon shape, and in the behavior code check if the owner object is a sprite, if it's, get the polygon from the collision shape from the sprite first animation and frame. Don't like it a lot.

Yes, ideally behaviors should 100% stay separated from the underlying object having them. We could provide way in the IDE to copy the collisions mask to physics polygon, but that's more a convenient option than what should be the default :/

**Comment by Lizard-13** [comment]
What I can do right now is enable a new string label in the behavior, here the user can enter the points in a list (as in the old IDE, but in pixels, no meters). Then try to parse it in the code.
It's ugly, but all the functionality will be there for the time it gets a nice editor :D

**Comment by Lizard-13** [comment]
Totally forgot the collision condition, now it works as the normal collision condition, i.e. both objects used in the condition are picked.
Also added the option to (don't) allow to sleep, and a condition to check if an object is sleeping, useful for these balance games.
collideandsleep

**Comment by blurymind** [comment]
Fantastic work here! 😸

**Comment by 4ian** [comment]
@Lizard-13 I can see that you've checked everything, do you think the extension is ready, or still things to do? :) Do you have the test bed that you shared as example?

**Comment by Lizard-13** [comment]
@4ian Yeah, almost ready, let me know if you think I'm forgetting something. I leaved it in standby because the polygon editor, but it will take some time so I guess we can release it as now, unless I've missed an old feature.
Oh, an important point, the behavior name. I've called it "Physik" so it doesn't conflict with the old behavior, but not sure if I leave it this way, rename it as "Physics2" or do you have another idea?

Let me update the two good examples, with comments and so.

**Comment by zatsme** [comment]
> Oh, an important point, the behavior name. I've called it "Physik" so it doesn't conflict with the old behavior, but not sure if I leave it this way, rename it as "Physics2" or do you have another idea?

As it was a different name It was relatively easy to run this alongside the 'old' physics and compare, which led me to thinking that more users should test new features before they were merged into the main editor.

I was thinking about this and wondered if it would be a good idea to have an 'Experimental/New Feature'  Editor or perhaps a checkbox (like Functions) so as to have experienced users testing new features/upgrades (electron etc) like this before releasing to the mainstream.

At least this will enable us to test it on Windows/Linux/Mac and Android/iPhone on various h/w and configurations to catch any major issues.

Zat. 🤔

**Comment by 4ian** [comment]
Physik looks cool, but as it is intended to be a replacement of the current Physics engine, I think Physics2 would be better suited. It's more obvious that it's "version 2" of the physics engine and that people should move to it. Also "Physik" could be considered as a typo and turn off some users if they see other typo in GDevelop.

> led me to thinking that more users should test new features before they were merged into the main editor.

Yes, I'd like to create "beta" releases of GDevelop and only update the "stable" release once in a while. Ideally that would be how it would work. But I need to first remove the "beta" version from GDevelop version number.
This is tracked here: https://trello.com/c/vMX7TOkv/130-remove-the-beta-from-the-gd5-version-number

A while ago there was quite a few feedbacks about things that should be fixed before doing this, but I think we're really close to be able to remove this "beta" version.
After all, I think most users don't notice they are using something that should be "beta", so we might drop the beta reference and start using it for versions with new extensions/risky changes from now.

**Comment by Lizard-13** [comment]
Renamed!

@zatsme You were using the extension already?, make a backup of the .json project file(s), edit it/them with a text editor and replace every "Physik" occurrence by "Physics2". I did it with my 3 example projects and everything is working as before.

**Comment by zatsme** [comment]
Yes, the bike example was to test the wheel joint 😁

**Comment by Lizard-13** [comment]
Also added the ability to manipulate the world time scale, I've to modify the action a bit so it doesn't run once per instance, just once. Will do it right now.

**Comment by Lizard-13** [comment]
Fixed!

Ok, I think it's good to go now, let me know if you find something weird.

Here are the final version of the test files:
The joints testbed: 
[Uploading Physics2JointsTest.zip…]()
The editable joints settings:
[Physics2JointsSettingsTest.zip](https://github.com/4ian/GDevelop/files/2655506/Physics2JointsSettingsTest.zip)

Also, I think I get the custom editor to work, just render a button but works.

**Comment by Lizard-13** [comment]
Should be ok now, sorry for that.

**Comment by Lizard-13** [comment]
@4ian Sorry for bother you again, but I get the custom editor to work in many ways and not sure what is the best design. In the behavior extension I pass a "dialog" property with the dialog name so the IDE knows which dialog to render, as you suggested.

The `PropertiesMapToSchema` defines a `getDialog` function so the `PropertiesEditor` knows there is a dialog property, it gets the dialog component from the `PropertiesRenderingService` and renders it.

Now the questions, should I keep the other properties getters and setters in the extension and pass the `schema` to the behavior dialog?, if I do it the other properties must not be rendered, I did it checking if there's a custom dialog in the properties, if there's one skip other "common" properties and render dialogs only. Then the behavior dialog can use the schema to read/write properties, using the getters/setters from the extension definition
Or the custom behavior dialog should handle the getter/setters?. if I do it I have to pass it the project, so the dialog can read/write the properties, I should do things like checking if the property is "boolean" or "number" every time, but this way the extension is abstracted from the IDE.

**Comment by 4ian** [comment]
That's an excellent question. I'll go for solution 2 as I think it's better to avoid skipping other properties or doing special work.

**Comment by Lizard-13** [comment]
The extension just passed this fire test:
zawarudo

**Comment by Lizard-13** [comment]
@4ian I think it's good to go now, added a "(beta)" in the behavior name to let the user know it isn't battle tested.

I will be busy this week, but then I'll continue to work in the editor to make a new PR, so far so good:
physics2editor
(That "Checky" checkbox is meant to be replaced by an image previewer)

**Comment by blurymind** [comment]
the layers/masks checkboxes really remind me of the ones in blender 2.7

**Comment by 4ian** [comment]
@Lizard-13 I'll check this PR this week-end :) Great job on this! Look forward to have it in GDevelop

**Comment by 4ian** [comment]
I've created a card for it on the roadmap by the way 😄 https://trello.com/c/sr8Rx4qe/214-improved-new-physics-behavior-with-joints-revolute-rope-motor-gear-new-shapes-masks-etc

File: Extensions/Physics2Behavior/JsExtension.js [hunk 1]
```diff
@@ -0,0 +1,2988 @@
...
+        t("Physics2 behavior (beta)"),
+        "Physics2Behavior",
+        t("Simulate physics, the successor of the old physics behavior"),
+        "",
+        "res/physics32.png",
+        "Physics2Behavior",
+        physics2Behavior,
+        sharedData
+      )
+      .setIncludeFile("Extensions/Physics2/physics2runtimebehavior.js")
```
Comment (4ian) [inline Extensions/Physics2Behavior/JsExtension.js]: Path should be `Extensions/Physics2Behavior` for all the include file no? I had to change them to get it to work :)

Comment (Lizard-13) [reply Extensions/Physics2Behavior/JsExtension.js]: Aha!, nice catch, when I renamed the folder removed the original one everywhere, maybe I've forgotten to remove it in the previews folder, loading the extension  anyway

Comment (4ian) [reply Extensions/Physics2Behavior/JsExtension.js]: Yeah, I think there was still the folder in the preview folder so you didn't see the issue :)

**Comment by 4ian** [comment]
I've tested, this is really cool! All joints are working well :) 
I think we'll be able to change the name of the behavior to Physics Engine 2.0 so that it's obvious to users that it's the new version.

**Comment by Lizard-13** [comment]
Sure, will fix these asap.

**Comment by Lizard-13** [comment]
Should I run prettier on the files?

**Comment by 4ian** [comment]
Yes let's run it on the files of the extension :) I plan at some point to run it on all extensions JS files and GDJS Runtime, so better if we start now.

Can you just make sure to run it with the same configuration at the one for newIDE (for consistency)? https://github.com/4ian/GDevelop/blob/master/newIDE/app/.prettierrc

**Comment by Lizard-13** [comment]
It's a bit worrying that I keep finding this kind of errors, so I'll read the full code again :mag:

**Comment by Lizard-13** [comment]
Added a small improvement (and fix) on joint removing, now it iterates over all the joints searching gears if the joint being removed is revolute or prismatic (the only ones that can be attached to gears).
Also reordered a property initialization, for consistency.

Couldn't find any other problem :+1:

**Comment by 4ian** [comment]
So this looks almost ready :) The editor is still for now showing the internal property names (in camelCase) which makes it a bit strange, and especially with masks/layers fields :)
Do you think we should add a way for the extension to provide "translations" for the property names? At least I could add something to automatically transform camelCase to Title Case. (My idea was property names would always be user friendly name, but it wasn't a great idea). 

Or do we wait for the editor that you've built? In this case we can still exclude advanced things while they are not ready.

**Comment by Lizard-13** [comment]
I think it's good to go, this way it's more obvious to the user it's in beta haha
Also, I'll work on the editor to get it ready as soon as possible, maybe for the beta next to the one you include this :)

While working in the editor, I can add an extra translatable name for the properties or convert the name to Title Case, what you like the most.

**Comment by 4ian** [comment]
I think we should still have an editor that does not look strange to new users, I'm pretty sure we'll get issues/bugs reports/confused users :)
It's fine to have the "full editor" coming later, even removing a few properties for now if needed, but let's still try to add extra name (would be the more flexible I think)

**Comment by Lizard-13** [comment]
Ok, extra parameter then :ok_hand:

**Comment by 4ian** [comment]
Sounds good! Will allow also later in other editors to have both a property name and a translation (user friendly name) for the user interface :)

**Comment by Lizard-13** [comment]
propertyname
Looking good, let me know if you want to change the property "name" in the .h file to something more clear as "displayName", also the map-to-schema returns the property name (e.g "fixedRotation") if the custom name hasn't been set, but if you want it could return the name converted to Title Case. Didn't do it yet because I've forgotten it and have problems with internet right now:(

**Comment by 4ian** [comment]
Looks like it works well :) May I suggest that we use "label" (Get/SetLabel in PropertyDescriptor.h and getLabel in PropertiesEditor and PropertiesMapToSchema.js)? 😇
displayName would have been fine, but: 1) label might be more suited/used than "display name" and 2) there is in fact already a "getLabel" function which is already supported in `PropertiesEditor`. It's used in InstancePropertiesEditor, so make sense to generalize this to all properties.

File: newIDE/app/src/PropertiesEditor/index.js [hunk 2]
```diff
@@ -206,7 +206,7 @@ export default class PropertiesEditor extends React.Component<Props, {||}> {
         key={field.name}
         fullWidth
         primary
-        label={field.getLabel(this.props.instances[0])}
```
Comment (4ian) [inline newIDE/app/src/PropertiesEditor/index.js]: There is already a "getLabel" in InstancePropertiesEditor, so we should not remove the usage of getLabel.

File: newIDE/app/src/PropertiesEditor/index.js [hunk 3]
```diff
@@ -84,7 +84,7 @@ export default class PropertiesEditor extends React.Component<Props, {||}> {
     if (field.valueType === 'boolean') {
       return (
         <InlineCheckbox
-          label={field.name}
+          label={field.getName(this.props.instances[0])}
```
Comment (4ian) [inline newIDE/app/src/PropertiesEditor/index.js]: So Flow typing of Field is very weak (it's just an Object) so Flow is not warning you but `getName` is not mandatory. For example, in InstancePropertiesEditor, it's not defined.

I suggest that we create a method, like _getFieldValue, called _getFieldLabel:

```
_getFieldValue(instances: Instances, field: Field): string {
  if (field.getLabel) return field.getLabel(instances[0]); //Maybe add a console.error if !instances[0], like in _getFieldValue. For some reason it happened according to some bug reports.

  return field.name;
}
```

File: newIDE/app/src/PropertiesEditor/index.js (hunk 2 above)
Comment (Lizard-13) [reply newIDE/app/src/PropertiesEditor/index.js]: Oooh, I didn't find the "getLabel" method in the PropertiesMapToSchema, so I thought the button code was old and broken... where is the getLabel definition?

Comment (Lizard-13) [reply newIDE/app/src/PropertiesEditor/index.js]: I see, there's a getLabel but defined in the instances properties only... so buttons won't work for generalized behavior properties, I think I tried to render buttons and it always crashed, hence the idea the code was outdated. I get it now :)

**Comment by 4ian** [comment]
Looks great :) Thanks for doing the changes.
I think we're almost ready to go. I'll add your demo/test beds as examples I think. I was also thinking of hiding the masks/layers while the new editor is not here, because I can already see people entering number instead of 1 and 0. What do you think? The way it works is not obvious if you're not familiar with Box2D :)

**Comment by Lizard-13** [comment]
You're so boring!, it's super cool when you write a "5" to convert the entire layer into a new and unpredictable number, let the user feel the "beta" :laughing:

Just kidding, the bits string is always converted into a valid 16 0/1 chars, and although I think the user will notice that, I know the results can be weird if you start writing other numbers or letters, will comment out the property asap :+1:

**Comment by zatsme** [comment]
The bike demo I made if you want to include it in the examples 😄 

[Bike Demo.zip](https://github.com/4ian/GDevelop/files/2698118/Bike.Demo.zip)

**Comment by Lizard-13** [comment]
Done, anyway you can change the layers/masks through events :)

**Comment by 4ian** [comment]
> Just kidding, the bits string is always converted into a valid 16 0/1 chars, and although I think the user will notice that, I know the results can be weird if you start writing other numbers or letters, will comment out the property asap 👍

Thanks for doing this 😄 

> The bike demo I made if you want to include it in the examples 😄

For sure! I'll add it :) 

Looks ready to go. I'll probably take a last look tonight and will merge it.

I'll surely make a release soon. Would be great if we can come up with an updated/improved wiki page about the Physics engine.
Maybe at least a note telling that an improved physics engine is now available as "Physics2" :) In the long term, I'll keep the help page for the old Physics extension accessible, but the help page of Physics2 will be the default one.

**Comment by Lizard-13** [comment]
Nooo, never added it! :open_mouth: 
Will try to add it now, otherwise you'll have to wait for the editor.

**Comment by zatsme** [comment]
One thought, if you have both engines in GD5 for a while you are bound to get users mixing up the engines and adding events from both!  Is there a way we can warn if this is happening?  Also, I think I mentioned this before about the two kinds of forces we can already add to objects, now there will be 3 different kinds! 😱

**Comment by Lizard-13** [comment]
@4ian Will start doing the polygon, merge it when you want, if I don't add the polygon now will do it for the editor, I'll rewrite the wiki page too!

**Comment by Lizard-13** [comment]
@zatsme The worst scenario is having two independent Box2D worlds around, objects from one world won't collide with the objects from the other, not so bad.
About the GD forces, what do you think @4ian ?

**Comment by 4ian** [comment]
@zatsme That's a very true concern!
I think people will be confused and will use both - so we should actively avoid this.

I think I'll work on something to warn about potential conflicting uses. Maybe gray out a behavior if another one is already used in the game (i.e: if Physics is used, gray out Physics2 and vice versa - including actions/conditions/expressions). If you still try to use them, you'll get a warning "You're trying to use XXX, but the game is already using YYY. Are you sure you want to use both? This could result in unexpected/conflicting behaviors of objects on the scene".

This could be extended to the use of forces with objects having a behavior like Physics/Physics2.

More generally, there are a few things that I'd like to warn users against (making weird hitboxes with platformer engine, using sprite of different sizes in platformer engine, launching music in a loop, and using image that are too large). All of these are not related, but could be really useful.

**Comment by Lizard-13** [comment]
> This could be extended to the use of forces with objects having a behavior like Physics/Physics2.

Pretty handy!

**Comment by 4ian** [comment]
> Pretty handy!

Yes, as much as I'd like to merge this now and have the new behavior and demo integrated, we should be a bit cautious - especially because GDevelop was criticized in the past for being confusing on multiple parts... It's also on us to make sure that things are easy to use properly, and hard to use improperly :)

And if we can avoid/reduce drastically the number of users asking the questions on the forum/discord, that's great :)
For example, in the same vein, if a game is not using any physics behavior, we should gray out Physics (the old one) and if the user clicks on it, he should get a message "You're about to use Physics, but Physics2 is recommended for new game as more complete and powerful. Are you sure you want to use it?". (We could even hide it!)
=> Instead of getting users asking for months "should I use physics or physics2", we'll get everybody to use the proper one :)

**Comment by zatsme** [comment]
Would it be an idea to have an extension/behaviour type Id? 

Then when a behaviour/e tension is used for the first time the ide can Check if another of the same type is already used and warn.

If physics was Id Phy1 and physics 2 was Phy2 then you can warn that a type of Phy was already in use and it's also a later/newer version.

That way if we ever get to a point where GD5 is being extended by many devs like construct and game maker then the parser can warn against issues automatically as long as the extension type is present. 

🤔

**Comment by 4ian** [comment]
Yes, that's what the internal name of the extensions are for :) 
In this case, the extension name is `Physics2` (and the name displayed to user is "Physics Engine 2.0" - even though the extension are never shown directly in GDevelop 5 for now).
The behavior name is `Physics2Behavior` ([see here](https://github.com/4ian/GDevelop/pull/775/files#diff-44c774e96a649c5ec44318372413a95dR360)) (and the name displayed to user is "Physics Engine 2.0 (beta)").

What could be done here is a quick helper in newIDE codebase that gives the extension used by a game (it's already done in `update-examples-information-from-resources-examples.js` in `computeUsedExtensions`, we should extract it in a separate file to use it in the IDE), then take the name of another extension (like "Physics2") and return a warning (or nothing if it's fine to use).
This would be using a list of "incompatibilities" (that could grow if more extensions are added in the future - even if ideally, extensions should behave properly).

This could be refined to events to detect usage of incompatibles actions/conditons on some objects (we should have all the tool needed to list the actions/conditions used on a specific object).

**Comment by 4ian** [comment]
For warning users about forces on object that have the physics behavior, that is surely a bit simpler that what I describe:

In the selector of instructions (internally, the word "instruction" represents a condition or an action), if the user choose a condition or action for an object, check the behavior of this object, and send it to the class/helper for incompatibilities.

I've added a task for it here: https://trello.com/c/xq91to0I/218-warn-the-user-against-using-incompatible-actions-conditions-like-using-force-on-an-object-that-have-the-physics-behavior

Here is the roadmap with all the "productivity" warnings that we could add: https://trello.com/b/qf0lM7k8/gdevelop-roadmap?menu=filter&filter=label:Productivity%20friendly%2Favoid%20common%20errors

If you have other ideas, let me know and I can add them here :)

**Comment by Lizard-13** [comment]
Should the polygon origin be at the center or top-left of the object?

**Comment by 4ian** [comment]
> Should the polygon origin be at the center or top-left of the object? 

What are the impacts/implications of both solutions?

**Comment by Lizard-13** [comment]
No idea, thinking about it... On object size change (and if shape scale is not changed), if the origin is at center it will stay centered (small offset in all the directions), but if the origin is at top-left all the shape offset will be at the bottom-right.

**Comment by zatsme** [comment]
Can't we define it as we do with sprites?

**Comment by Lizard-13** [comment]
A choice Polygon Origin: TopLeft/Center ?, could be, don't know.

**Comment by 4ian** [comment]
If we can, we should make this choice for the user no?

> Can't we define it as we do with sprites?

I guess the way polygons are defined in Box2D is a bit different than hitboxes (there can be a single polygon for example). Also, polygon can't/should not change for every frame - this would be a nightmare in terms of performance.

**Comment by Lizard-13** [comment]
> If we can, we should make this choice for the user no?

I'm checking out GD4 (oof haha), and the polygon editor indeed let the user choose object center and origin (and for some reason top-left option is commented out in the sources)

> Also, polygon can't/should not change for every frame - this would be a nightmare in terms of performance.

For sure, what I'm doing right now is storing the vertices in a memory buffer, they aren't recomputed on shape size change, but you've to pass them to the body again in a new shape, so every contact has to be recomputed.

**Comment by Lizard-13** [comment]
Good and bad news, good ones: I get it to work, and have proofs :)
scale1
scale2
The smaller and darker shape is the expected size at 0.5 shape scale, as you can see the triangle scales correctly on center, and the green one scales on top-left.

The bad ones, @4ian , do you have any idea why the parser converts an empty array into a JSON empty object, and non-empty arrays break the parser, skipping next properties?
For example I add the default property `vertices: [],` around here: https://github.com/4ian/GDevelop/pull/775/files#diff-44c774e96a649c5ec44318372413a95dR277
and even edit the `game.json` file to add the vertices array manually, but that doesn't matter, on preview/export GD replaces them with broken objects. `vertices: []` is coverted to `"vertices": {}`, and something like `vertices: [1,2,3]` into `"vertices": {"": {"3": "shape"}}" or something like that, mixing the next property and skipping the others.
The raw JSON content (https://github.com/4ian/GDevelop.js/blob/234ecd8e03c3aa7ab62095907bcad9e767878908/Bindings/BehaviorJsImplementation.h#L31) is correctly saved.
I get it to work and take the screenshots editing the `data.js` file.

**Comment by Lizard-13** [comment]
@4ian Merge it if you want, the conversation is getting long and it seems it will take ma a while to get what's going on, for a feature that won't be available anyway :)

**Comment by 4ian** [comment]
In short it's because the Serializer was originally created to be compatible with XML and JSON. XML does not have arrays but multiple children with the same name. 
The `SerializerElement`s that are created by the gd::Serializer don't contain the information that this thing is an array (because XML don't have this). So I made a method `ConsiderAsArrayOf` that you can call on SerializerElement, to be able to iterate on the children (not useful for JSON, but was useful for XML). It's manually called by the methods that serialize/unserialize. As all we're doing here is parsing and writing back JSON, without calling `ConsiderAsArrayOf`, the notion that this thing was an array is lost.

I think we should be able to have a "isArray" boolean stored in the SerializerElement when reading the JSON. Then when writing back the SerializerElement, it would be properly outputted as an array. 
(This would be an extra property in SerializerElement, the property arrayOf, set by ConsiderAsArrayOf, would still be there).

I'll take a look!

**Comment by 4ian** [comment]
@Lizard-13 Fix was actually quite simple. Made the fix in 238b6a2a092f38e0f572a3872115fd1708caa5b2 and added tests for this.

I've updated libGD.js, remove it from public/libGD.js and relaunch `npm start`/`yarn start` (did not bumped the version as it's a bugfix).

Let me know if you still have an issue!

**Comment by Lizard-13** [comment]
I'm a bit lost with this one, I've cherry picked your commits into my ligGD.js, but still failing to read the array. the SerializerElement passed here:
https://github.com/4ian/GDevelop.js/blob/be3e0b8d63fecfb4a4aa90e01e8a9b4f2dbd966c/Bindings/BehaviorJsImplementation.cpp#L83
Is not considering the "vertices" child as an array, so I guess the problem comes from the project unserialization. Will continue the debug later and start the pull request for the polygon and editor after you merge it, to ensure there are no further modifications and conflicts.
*Maybe I'm doing something wrong, but I guess opening the game file should read the array correctly no matter the behavior code.

**Comment by 4ian** [comment]
If you have merged master into your branch, there is no need to cherry pick anything right? 
You have https://github.com/4ian/GDevelop/commit/238b6a2a092f38e0f572a3872115fd1708caa5b2 in your commits right?

**Comment by Lizard-13** [comment]
Hey, you're right, but I did it before merging master, not sure why :sweat_smile: 
Yeah, I've the modifications and recompiled libGD.js myself (so I can debug it around), the tests you've added seems pretty solid, so I must be doing a mistake somewhere... but as said before, lets continue it in a new pull request, it's going off topic for the current one :)

**Comment by 4ian** [comment]
Let's merge this now! I'll hide the extension for the next version and add a preference to display it once we have help and a way to make sure users can't misuse the two behaviors ;)

**Comment by 4ian** [comment]
Huge huge thanks for your totally awesome work on this. 💯🎉 This is a huuuge feature! I have no doubt that lots of game will use it soon - and the demos are super impressive. 😎

It's a great example of GDevelop being able to handle complex games with advanced physics! 🚀
Thanks again and keep up the good job!

**Comment by jjhesk** [comment]
Does the polygon physics work now on GD5? Can show me some example?

**Comment by 4ian** [comment]
Please ask this on the forum or on the Discord :)

**Pull request merged**